lib/puppet/type/acl.rb

```diff
@@ -226,6 +226,9 @@ def initialize(*args)
          to the manifest.
          Reference: #{value.inspect}"
       end
+      if !(value['type'].nil? || value['type'].empty?)
+        warning('Permission `type` is deprecated and has been replaced with perm_type for allow or deny')
+      end
```

Comment:
Validation should sit within the validation, and warnings should reside with that context instead of in the initialization.  We would be spinning additional cycles to validate this even when getting the instances back and creating the objects.